firebaseai/GroundingSample/Views/GroundedResponseViews.swift

```diff
@@ -0,0 +1,148 @@
+        .clipShape(RoundedRectangle(cornerRadius: 20, style: .continuous))
+    }
+    // Set a max width to prevent the bubble from taking the full screen width.
+    .frame(maxWidth: UIScreen.main.bounds.width * 0.8, alignment: .trailing)
+  }
+}
+
+/// The content (text and sources) that goes inside the model's response bubble.
+struct ModelResponseContentView: View {
+  let text: String
+  let groundingMetadata: GroundingMetadata?
+
+  var body: some View {
+    VStack(alignment: .leading, spacing: 12) {
+      Text(text)
+
+      if let groundingChunks = groundingMetadata?.groundingChunks, !groundingChunks.isEmpty {
+        Divider()
+        VStack(alignment: .leading) {
+          Text("Sources")
+            .font(.footnote)
+            .foregroundColor(.secondary)
+
+          ForEach(0 ..< groundingChunks.count, id: \.self) { index in
+            if let webChunk = groundingChunks[index].web {
+              SourceLinkView(
+                title: webChunk.title ?? "Untitled Source",
+                uri: webChunk.uri
+              )
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+/// The complete visual component for the model's turn. It handles compliance checks internally.
+struct ModelResponseTurnView: View {
+  let response: GenerateContentResponse
+
+  var body: some View {
+    // Use `if-let` to safely unwrap the first candidate. This is the correct SwiftUI pattern.
+    if let candidate = response.candidates.first {
+      // A response is non-compliant ONLY if groundingMetadata exists but searchEntryPoint is nil.
+      let isNonCompliant = (candidate.groundingMetadata != nil && candidate.groundingMetadata?
+        .searchEntryPoint == nil)
+
+      if isNonCompliant {
+        ComplianceErrorView()
+      } else {
+        // This branch handles both compliant grounded responses and non-grounded responses.
+        HStack(alignment: .top, spacing: 8) {
+          VStack(alignment: .leading, spacing: 8) {
+            ModelResponseContentView(
+              // Use the convenience accessor on the response object.
+              text: response.text ?? "No text in response.",
+              groundingMetadata: candidate.groundingMetadata
+            )
+            .padding(12)
+            .background(Color.modelResponseBackground)
+            .clipShape(RoundedRectangle(cornerRadius: 20, style: .continuous))
+
+            if let searchEntryPoint = candidate.groundingMetadata?.searchEntryPoint {
+              WebView(htmlString: searchEntryPoint.renderedContent)
+                .frame(height: 44)
+                .clipShape(RoundedRectangle(cornerRadius: 22))
+            }
+          }
+        }
+        .frame(maxWidth: UIScreen.main.bounds.width * 0.8, alignment: .leading)
+      }
+    } else {
+      // This `else` branch handles the case where the response has no candidates.
+      ComplianceErrorView(
+        message: "The response was blocked or contained no content."
+      )
+    }
+  }
+}
```

Comment:
Could we make some minor edits to the existing [`MessageView`](https://github.com/firebase/quickstart-ios/blob/86d7d4edf760368de8079a5e16723bb376215cc7/firebaseai/ChatExample/Views/MessageView.swift#L68-L93) instead of brand new views for this screen?

Reply:
I refactored all of this to use the existing `ConversationScreen` and `MessageView`. Let me know what you think!

Reply:
Thanks, Daniel! I think this will be much easier to maintain! I've got some minor nits that I'll share on Monday and I hope you don't mind that I'll push a fix (🤞) for the `project.pbxproj` to _hopefully_ get it building on CI. We'll still need to swap it back to use SemVer (version `12.0.0`) before merging but will need to wait for `firebase-ios-sdk` to get tagged for that.